action/protocol/staking/protocol.go

```diff
@@ -114,3 +119,36 @@ func (p *Protocol) handleRestake(ctx context.Context, act action.Action, sm prot
 	// TODO
 	return nil, nil
 }
+
+func (p *Protocol) hasDelegate(name address.Address) bool {
+	return p.delegates.Contains(name)
+}
+
+func (p *Protocol) addDelegate(d *Delegate) error {
+	if err := d.Validate(); err != nil {
+		return err
+	}
+
+	if !p.hasDelegate(d.Owner) {
+		p.delegates[d.Owner.String()] = d
+	}
+	return nil
```

Comment:
is it possible this function be called multiple times with the same key? if so, should we return error since we don't overwrite? if not, why we test "hasDelegate"

Reply:
good point, thx

need to think the correct logic, since we discussed ysterday that we'll use 12-byte name. In this case, if an active delegate change the name, then we need to update this delegate, even though it's already in the map